### DISCUSSION TITLE
graphql v16 support?

### DISCUSSION THREAD

**Discussion by jayalfredprufrock** [tinacms/tinacms#3486] [General]
I have a pretty solid proof of concept for Gatsby + Tina that works great. But because tinacms is on GraphQL v15 and Gatsby 5 is on v16, I'm forced to stay on Gatsby v4 since the GraphQL lib is super strict about not letting multiple version coexist. Any plans on getting Tina on a more recent version of the GraphQL library? Or perhaps somebody can share a workaround?

**Comment by jayalfredprufrock** [comment]
To clarify, when forcing Tina to use graphql v16 with package.json overrides, the error I get looks like this:

```
visitFn.call is not a function
/Users/Play/Code/modern-doc/tools/gatsby-starter-modern-doc/node_modules/graphql/language/visitor.js:197
          : visitFn.call(visitor, node, key, parent, path, ancestors);
                    ^

TypeError: visitFn.call is not a function
    at visit (/Users/Play/Code/modern-doc/tools/gatsby-starter-modern-doc/node_modules/graphql/language/visitor.js:197:21)
    at Object.plugin (/Users/Play/Code/modern-doc/tools/gatsby-starter-modern-doc/node_modules/@tinacms/cli/dist/index.js:1015:51)
    at /Users/Play/Code/modern-doc/tools/gatsby-starter-modern-doc/node_modules/@graphql-codegen/core/cjs/execute-plugin.js:36:54
    at /Users/Play/Code/modern-doc/tools/gatsby-starter-modern-doc/node_modules/@graphql-codegen/plugin-helpers/cjs/profiler.js:7:49
```
Apparently several libraries broke in a similar way. Here is an example: https://github.com/croutonn/graphql-codegen-plugin-typescript-swr/commit/50b8228e60a077906994b2aa83fc9b33ce3b3fb7

Most were fixed by using `oldVisit` from `@graphql-codegen/plugin-helpers` instead of the `visit` exported from `graphql`.

I applied the same fix in a temporary repo here to unblock myself: https://github.com/jayalfredprufrock/tina-cms-cli-graphql-16
It's not properly tracking the tinacms monorepo so staying in sync will be a pain. Hoping their is an official solution on the way or confirmation that a PR that applies the above fix would get merged. Its unclear to me whether such a change would require a major version bump.

**Reply by logan-anderson** [reply]
@jayalfredprufrock we are open to contributions if you want to make a PR to the [tinacms repo](https://github.com/tinacms/tinacms). I tried taking a look at your example but the `src` folder is not there (only the compiled code).

**Reply by smiley-uriux** [reply]
I'm happy to put together a PR, just wanted to make sure there was some interest and/or no immediate blockers. It was literally a one line change so it was easier to modify the compiled code directly and push it up rather than compile it from source and publish it to NPM . Somebody with more knowledge of the ecosystem will have to comment on what such a change means, i.e. semver. It feels like this would be a breaking change and require a major version bump but I'm not positive.

Ideally libraries like Gatsby and Tina would be using peer dependencies for graphql, however I realize both projects rely heavily on type generation so I can see why they've opted to include them directly. Hopefully the community is working on a better solution.

**Reply by logan-anderson** [reply]
@smiley-uriux I can't think of any reason why there would be blockers.  A PR would be more than welcome.  This could probably be a minor or patch bump since it would be backwards compatible.

**Reply by smiley-uriux** [reply]
@logan-anderson I believe such a change would require graphql to be on v16 _for all dependencies_ within a project. If somebody is pairing Tina with some other project that carries graphql v15 as a normal dependency, that would break things for them. Again, not sure how common that is...but seeing as graphql v17 is already in alpha, I imagine v16 to be more common in the wild right now anyways.  Either way I'll put together a PR soon so the team can review. I appreciate the dialogue!

**Reply by logan-anderson** [reply]
That makes sense and I agree 16 is probably more common. Most of our users are using Next.js or an SSG framework. The only one that I know of that also has GraphQL as a dependency is Gatsby.  Since Gatsby is on version 16 I think it would make the most sense for us to be on that version as well. We will talk this over internally but I think a Minor bump makes the most sense. I think it would be backwards compatible in most cases. 

Thanks again

**Reply by jayalfredprufrock** [reply]
https://github.com/tinacms/tinacms/pull/3507

**Comment by logan-anderson** [comment]
@jayalfredprufrock Do you get any errors from Gatsby when forcing the graphQL version to 15?

**Reply by smiley-uriux** [reply]
@logan-anderson I sure did, that was the first thing I tried. Gatsby introduced auto-generated graphql types few versions ago though so it relies heavily on the graphql + codegen ecosystem and downgrading seemed to create more havoc than the upgrade path. I also explored using Gatsby v4 which uses graphql v15, but then I had issues with MDX handling since there was a major change to MDX from v4-v5 (upgraded to MDX v2). 

Since implementing the fix I mentioned above, I haven't experienced any issues with Tina or Gatsby and all appears to work great. I'm incredibly impressed with the integration and am very excited to work with Tina again. I was pretty bummed when support got dropped for NextJs but now I can see it was worth the wait!

**Comment by gsicvj** [comment]
It's so wrong that a CMS becomes a dependency bottleneck of a project, when it's main role is purely content management.
Local preview isn't a good tradeoff in my opinion.

**Reply by logan-anderson** [reply]
I am not 100% sure what you mean? Mind providing a bit more detail?

**Reply by gsicvj** [reply]
This is, as the author mentions, about using Tina in a Gatsby project.

As far as I read the documentation on Tina, this is my pain point:

> The reason we want to run the GraphQL API with our site is so that:
When our static pages build in CI, they can source their content from the local files using the GraphQL API.
In development, we can test out Tina with our local files.
Source: https://tina.io/docs/graphql/cli/#running-the-graphql-api

There are two potential use cases I see with a graphql problem.
The first case is using forestry with old dependencies and having to migrate to Tina.
The second case is having up to date dependencies and using Tina with older versions of graphql. 

The build seems to fail when Gatsby and Tina graphql versions differ, it would be nice if it didn't.
Maybe the build errors are an unexpected outcome, but I wouldn't want to care about which graphql version is used by Tina for query previews.

All the graphql query previews are already accessible via gatsby's /__graphql_ suffix.
Maybe for performance reasons Tina also uses graphql for mutations and writing data internally.

If not, having this in a Gatsby project with Tina seems redundant, especially when Forestry was working just fine.
And in any case, it doesn't feel right that a CMS is opinionated so much, that it tells you which graphql dependency it needs.

Please let me know if I'm missing something. I'd be happy to discuss this further.
